osquery/tables/networking/windows/routes.cpp

```diff
@@ -0,0 +1,168 @@
+/*
+*  Copyright (c) 2014-present, Facebook, Inc.
+*  All rights reserved.
+*
+*  This source code is licensed under the BSD-style license found in the
+*  LICENSE file in the root directory of this source tree. An additional grant
+*  of patent rights can be found in the PATENTS file in the same directory.
+*
+*/
+
+#define WIN32_LEAN_AND_MEAN
+
+#include <memory>
+#include <string>
+#include <windows.h>
+#include <winsock2.h>
+
+#include <wS2tcpip.h>
+#include <ws2ipdef.h>
+
+#include <iphlpapi.h>
+#include <mstcpip.h>
+
+#include <boost/algorithm/string/join.hpp>
+#include <osquery/tables.h>
+
+#include "osquery/core/conversions.h"
+#include "osquery/core/windows/wmi.h"
+
+namespace osquery {
+namespace tables {
+
+std::map<unsigned long, std::shared_ptr<IP_ADAPTER_INFO>>
+getAdapterAddressMapping() {
+  std::map<unsigned long, std::shared_ptr<IP_ADAPTER_INFO>> returnMapping;
+  auto dwBufLen = 0UL;
+  auto dwStatus = GetAdaptersInfo(NULL, &dwBufLen);
+
+  if (dwStatus == ERROR_BUFFER_OVERFLOW) {
+    std::vector<BYTE> buffer(dwBufLen);
+    auto pAdapterInfo = reinterpret_cast<PIP_ADAPTER_INFO>(buffer.data());
+    dwStatus = GetAdaptersInfo(pAdapterInfo, &dwBufLen);
+
+    if (dwStatus != NO_ERROR) {
+      return returnMapping;
+    }
+
+    while (pAdapterInfo) {
+      auto sharedPtr = std::make_shared<IP_ADAPTER_INFO>(*pAdapterInfo);
+      returnMapping.insert(std::make_pair(pAdapterInfo->Index, sharedPtr));
+      pAdapterInfo = pAdapterInfo->Next;
+    }
+  }
+
+  return returnMapping;
+}
+
+PMIB_IPINTERFACE_TABLE getInterfaces(int type = AF_UNSPEC) {
+  auto interfaceTable = std::make_unique<PMIB_IPINTERFACE_TABLE>();
+
+  auto dwRetVal = GetIpInterfaceTable(type, interfaceTable.get());
+  if (dwRetVal != NO_ERROR) {
+    return nullptr;
+  }
+
+  return *interfaceTable.get();
+}
+
+std::map<unsigned long, MIB_IPINTERFACE_ROW> getInterfaceRowMapping(
+    int type = AF_UNSPEC) {
+  std::map<unsigned long, MIB_IPINTERFACE_ROW> returnMapping;
+  PMIB_IPINTERFACE_TABLE interfaces;
+
+  if ((interfaces = getInterfaces(type)) != nullptr) {
+    for (unsigned long i = 0; i < interfaces->NumEntries; ++i) {
+      MIB_IPINTERFACE_ROW currentRow = interfaces->Table[i];
```

Comment:
You're invoking the copy-assignment here, if the `MIB_IPINTERFACE_ROW` content is large, this is inefficient. The code will again copy the content into the `std::pair<T, U>` If you store a reference `const auto& currentRow = interfaces->Table[i]` the content will only be copied once.